Adding minimal MPI executor

hkaiser opened the pull request:
- renaming `hpx::mpi::async` --> `hpx::mpi::detail::async`
- adding another test

weilewei commented on libs/mpi/tests/unit/mpi_ring_async_executor.cpp:
```diff
@@ -0,0 +1,165 @@
+    hpx::mpi::executor exec(MPI_COMM_WORLD);
+
+    unsigned int const n_loops = 20;
+    std::atomic<int> counter = n_loops;
```
If I change the way of using construction, like, `std::atomic<int> counter(n_loops)`, I would not receive compile error. Otherwise, I have a compilaion error:
```
[weile@login3.summit build_Debug]$ make tests.unit.modules.mpi.mpi_ring_async_executor
[  0%] Built target hpx_preprocessor
[  6%] Built target hpx_config
[  6%] Built target hpx_assertion
[  6%] Built target hpx_type_support
[ 12%] Built target hpx_concepts
[ 12%] Built target hpx_filesystem
[ 12%] Built target hpx_format
[ 12%] Built target hpx_timing
[ 18%] Built target hpx_logging
[ 18%] Built target hpx_thread_support
[ 25%] Built target hpx_errors
[ 25%] Built target hpx_basic_execution
[ 31%] Built target hpx_hashing
[ 31%] Built target hpx_concurrency
[ 31%] Built target hpx_datastructures
[ 31%] Built target hpx_debugging
[ 31%] Built target hpx_functional
[ 31%] Built target hpx_naming_base
[ 31%] Built target hpx_serialization
[ 31%] Built target hpx_iterator_support
[ 37%] Built target hpx_util
[ 37%] Built target hpx_topology
[ 37%] Built target hpx_affinity
[ 37%] Built target hpx_allocator_support
[ 37%] Built target hpx_cache
[ 37%] Built target hpx_coroutines
[ 37%] Built target hpx_hardware
[ 37%] Built target hpx_memory
[ 37%] Built target hpx_plugin
[ 43%] Built target hpx_program_options
[ 50%] Built target hpx_threading_base
[ 50%] Built target hpx_synchronization
[ 50%] Built target hpx_threadmanager
[ 50%] Built target hpx_resource_partitioner
[ 50%] Built target hpx_local_lcos
[ 50%] Built target hpx_execution
[ 50%] Built target hpx_statistics
[ 50%] Built target hpx_algorithms
[ 56%] Built target hpx_batch_environments
[ 56%] Built target hpx_checkpoint
[ 62%] Built target hpx_static_reinit
[ 62%] Built target hpx_collectives
[ 62%] Built target hpx_compute
[ 62%] Built target hpx_compute_cuda
[ 62%] Built target hpx_mpi
[ 62%] Built target hpx_resiliency
[ 62%] Built target hpx_segmented_algorithms
[ 62%] Built target hpx_testing
[106%] Built target hpx
[106%] Built target hpx_wrap
[106%] Built target hpx_init
Scanning dependencies of target mpi_ring_async_executor_test
[106%] Building CXX object libs/mpi/tests/unit/CMakeFiles/mpi_ring_async_executor_test.dir/mpi_ring_async_executor.cpp.o
/gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/libs/mpi/tests/unit/mpi_ring_async_executor.cpp: In function 'int hpx_main(boost::program_options::variables_map&)':
/gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/libs/mpi/tests/unit/mpi_ring_async_executor.cpp:82:32: error: use of deleted function 'std::atomic<int>::atomic(const std::atomic<int>&)'
     std::atomic<int> counter = n_loops;
                                ^~~~~~~
In file included from /gpfs/alpine/proj-shared/cph102/weile/dev/install/boost/1.68.0/include/boost/smart_ptr/detail/spinlock_std_atomic.hpp:19,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/install/boost/1.68.0/include/boost/smart_ptr/detail/spinlock.hpp:47,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/libs/errors/include/hpx/errors/exception_list.hpp:15,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/build_Debug/libs/errors/include/hpx/errors.hpp:16,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/hpx/runtime/agas/interface.hpp:15,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/hpx/runtime/applier/apply.hpp:14,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/hpx/apply.hpp:11,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/hpx/hpx.hpp:10,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/libs/mpi/tests/unit/mpi_ring_async_executor.cpp:7:
/autofs/nccs-svm1_sw/summit/gcc/8.1.1/include/c++/8.1.1/atomic:668:7: note: declared here
       atomic(const atomic&) = delete;
       ^~~~~~
/autofs/nccs-svm1_sw/summit/gcc/8.1.1/include/c++/8.1.1/atomic:672:17: note:   after user-defined conversion: 'constexpr std::atomic<int>::atomic(std::atomic<int>::__integral_type)'
       constexpr atomic(__integral_type __i) noexcept : __base_type(__i) { }
                 ^~~~~~
make[3]: *** [libs/mpi/tests/unit/CMakeFiles/mpi_ring_async_executor_test.dir/mpi_ring_async_executor.cpp.o] Error 1
make[2]: *** [libs/mpi/tests/unit/CMakeFiles/mpi_ring_async_executor_test.dir/all] Error 2
make[1]: *** [libs/mpi/tests/unit/CMakeFiles/tests.unit.modules.mpi.mpi_ring_async_executor.dir/rule] Error 2
make: *** [tests.unit.modules.mpi.mpi_ring_async_executor] Error 2
```

hkaiser replied:
@weilewei this should be ok now.

hkaiser commented:
@biddisco I added some more changes to this PR, please review again. Also, I think that the underlying PR (#4277) needs more work before it can go in. Mainly:

1) The functions `hpx::mpi::init`/`hpx::mpi::finalize` can be potentially used to perform incompatible operations: 
- initialize/finalize MPI itself (which most probably should happen outside of HPX)
- initialize/finalize the scheduler mode and the polling registration (which most probably should happen from an HPX thread.
I think we should completely remove the invocation of `MPI_init` and `MPI_finalize` from this code. We can rely on MPI being initialized by the user in  any case. A simple assert that makes sure MPI was initialized would be sufficient.

2) Both functions take a `pool_name` argument which is way to easy to mix (i.e. pass different values to init/finalize). For the scheduler mode, I have added an RAII type `enable_scheduler_polling` that circumvents this issue.

3) Either fully implement or remove the MPI error handler functionality.

For now, I have added some default values to the arguments that should help simplifying the use of this infrastructure.

weilewei commented:
I got the following error from this line: https://github.com/STEllAR-GROUP/hpx/blob/c49e9aebc4a6b68687d78e1ed8e281d6e5468611/libs/mpi/include/hpx/mpi/mpi_future.hpp#L126

```
Scanning dependencies of target hpx_G2_ring
[ 50%] Building CXX object hpx_mpi_futures/CMakeFiles/hpx_G2_ring.dir/hpx_G2_ring.cpp.o
In file included from /gpfs/alpine/proj-shared/cph102/weile/dev/install/hpx_mpi_async_Debug/include/hpx/mpi/mpi_executor.hpp:13,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/install/hpx_mpi_async_Debug/include/hpx/mpi.hpp:12,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/Ring_example_MPI_CUDA/hpx_mpi_futures/hpx_G2_ring.cpp:3:
/gpfs/alpine/proj-shared/cph102/weile/dev/install/hpx_mpi_async_Debug/include/hpx/mpi/mpi_future.hpp:126:28: error: 'moodycamel' does not name a type
         using queue_type = moodycamel::ConcurrentQueue<future_data_ptr>;
                            ^~~~~~~~~~
/gpfs/alpine/proj-shared/cph102/weile/dev/install/hpx_mpi_async_Debug/include/hpx/mpi/mpi_future.hpp:127:16: error: 'queue_type' does not name a type; did you mean 'decltype'?
         static queue_type request_queue_;
                ^~~~~~~~~~
                decltype
/gpfs/alpine/proj-shared/cph102/weile/dev/install/hpx_mpi_async_Debug/include/hpx/mpi/mpi_future.hpp: In function 'void hpx::mpi::detail::add_to_request_queue(hpx::mpi::detail::future_data_ptr)':
/gpfs/alpine/proj-shared/cph102/weile/dev/install/hpx_mpi_async_Debug/include/hpx/mpi/mpi_future.hpp:136:13: error: 'request_queue_' was not declared in this scope
             request_queue_.enqueue(data);
             ^~~~~~~~~~~~~~
/gpfs/alpine/proj-shared/cph102/weile/dev/install/hpx_mpi_async_Debug/include/hpx/mpi/mpi_future.hpp:136:13: note: suggested alternative: 'add_to_request_queue'
             request_queue_.enqueue(data);
             ^~~~~~~~~~~~~~
             add_to_request_queue
/gpfs/alpine/proj-shared/cph102/weile/dev/install/hpx_mpi_async_Debug/include/hpx/mpi/mpi_future.hpp: In function 'void hpx::mpi::poll()':
/gpfs/alpine/proj-shared/cph102/weile/dev/install/hpx_mpi_async_Debug/include/hpx/mpi/mpi_future.hpp:246:24: error: 'request_queue_' is not a member of 'hpx::mpi::detail'
         while (detail::request_queue_.try_dequeue(val))
                        ^~~~~~~~~~~~~~
/gpfs/alpine/proj-shared/cph102/weile/dev/install/hpx_mpi_async_Debug/include/hpx/mpi/mpi_future.hpp:246:24: note: suggested alternative: 'add_to_request_queue'
         while (detail::request_queue_.try_dequeue(val))
                        ^~~~~~~~~~~~~~
                        add_to_request_queue
make[2]: *** [hpx_mpi_futures/CMakeFiles/hpx_G2_ring.dir/hpx_G2_ring.cpp.o] Error 1
make[1]: *** [hpx_mpi_futures/CMakeFiles/hpx_G2_ring.dir/all] Error 2
make: *** [all] Error 2
```

hkaiser commented:
@weilewei sorry, I removed an header assuming it's not needed. That should be fine now.

weilewei commented:
> @weilewei sorry, I removed an header assuming it's not needed. That should be fine now.

Now it compiles and I can run simple send-and-receive program. I will run my ring algorithm (+gpuDirect) tomorrow and see how it goes. I will keep you posted. No worry, and thanks!

weilewei commented:
> > @weilewei sorry, I removed an header assuming it's not needed. That should be fine now.
> 
> Now it compiles and I can run simple send-and-receive program. I will run my ring algorithm (+gpuDirect) tomorrow and see how it goes. I will keep you posted. No worry, and thanks!

@hkaiser I can verify that this PR works fine for my ring algorithm (with gpuDiect) meaning that with hpx mpi executor, cuda array can be sent and received through Nvidia NVlink (attached to mpi interface) on Summit.

biddisco commented:
Thanks @weilewei That's great. One of my main motives for pushing this is because I want it in DCA++ and other codes like it. I will give your GPU direct stuff a try too.

biddisco commented:
PS. I will clean this PR up a bit over the weeken - or when I'm back at work on Monday and try to get it merged ASAP.

biddisco commented:
the executor version of the test hangs, so I'll need to make some tweaks

biddisco commented:
Our test application also locks up using this new version with executors.

hkaiser commented:
> the executor version of the test hangs, so I'll need to make some tweaks

Ok, thanks! What did I break?

hkaiser commented:
> Our test application also locks up using this new version with executors.

I don't think it's a problem introduced by the executor. This one is very much a thin wrapper around the code that was there. I must have broken something in the implementation, sorry.
